File: app/src/main/java/java238/background/RobotProject.java
/**
 * loads the frc robot project and exposes methods to get basic info
 */

package java238.background;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.MapperFeature;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Scanner;
import java.util.regex.Pattern;

public class RobotProject {

    private String rootDirectory;
    private ObjectMapper mapper;
    private String jsonString;
    private static AmodeList amodesList;
    private static List<CommandInfo> commands = new ArrayList<>();
    private ArrayList<String> trajectories;

    public RobotProject(String rootDirectory) {
        this.rootDirectory = rootDirectory;
        mapper = new ObjectMapper();
        mapper.configure(MapperFeature.ACCEPT_CASE_INSENSITIVE_PROPERTIES, true);
    }

    public static OSName getOSName() {
            return System.getProperty("os.name").toLowerCase().contains("win") ? OSName.Windows : OSName.Unix;
    }

    public void indexCommands() throws FileNotFoundException {
        Pattern pattern = Pattern.compile(Constants.autoModeAnnotationRegex);

        File dir = new File(rootDirectory + Constants.commandDirectory);
        File[] directoryListing = dir.listFiles();

        if (directoryListing == null)  return;

        for (File child : directoryListing) {
            CommandInfo info;
            String[] params;
            Scanner scanner = new Scanner(child);
            var line = scanner.nextLine();

            // Do something with child
            while (scanner.hasNextLine()) {
                line = scanner.nextLine();
                var findPattern = scanner.findInLine(pattern);

                if (findPattern == null) continue;

                params = findPattern.substring(44, findPattern.length() - 2).split(", ");

                ArrayList<String> paramsList = new ArrayList<>(params.length);
                paramsList.addAll(Arrays.asList(params));

                info = new CommandInfo(child.getName().substring(0, child.getName().length() - 5), paramsList);
                commands.add(info);

                System.out.println(info.getName() + " with parameters " + info.getParameters().toString().strip());
                break;

            }
        }
    }
    public void indexTrajectories() {
        trajectories = new ArrayList<>();
        Pattern pattern = Pattern.compile(rootDirectory + Constants.deployDirectory + Constants.pathPlannerDir);
        File dir = new File(rootDirectory + Constants.deployDirectory + Constants.pathPlannerDir);

        File[] dirList = dir.listFiles();
        assert dirList != null;
        for (File child : dirList) {
            System.out.println(child.getName());
            trajectories.add(child.getName().split("\\.path")[0]);
        }
        System.out.println(trajectories);
    }

    public ArrayList<String> getTrajectories() {
        return trajectories;
    }



    public RobotProject() {
        mapper = new ObjectMapper();
        mapper.configure(MapperFeature.ACCEPT_CASE_INSENSITIVE_PROPERTIES, true);
    }

    public void setRootDirectory(String rootDirectory) {
        this.rootDirectory = rootDirectory;
    }

    public void loadRobotProject() {
        try {
            jsonString = Files.readString(Paths.get(rootDirectory + Constants.deployDirectory + Constants.amodeFile));
        } catch (IOException e) {
            throw new RuntimeException(e);
        }

    }

    public String getJsonString() {
        return jsonString;
    }

    public List<CommandInfo> getCommands() {
        return commands;
    }

    public AmodeList getAutoModes() {
        if (amodesList != null) {
            return amodesList;
        } else {
            try {
                amodesList = mapper.readValue(jsonString, new TypeReference<AmodeList>() {});
            } catch (JsonProcessingException e) {
                throw new RuntimeException(e);
            }
            return amodesList;
        }
    }

    public void setAutoModes(AmodeList list) {
        amodesList = list;
        saveOut();
    }

    public void saveOut() {
        try {
            System.out.println(mapper.writeValueAsString(amodesList));
        } catch (JsonProcessingException e) {
            throw new RuntimeException(e);
        }
    }

    public void tempSaveDraggedCommand () {

    }




    public static class Constants {
        public static final String amodeFile = "amode238.txt";
        public static final String deployDirectory = getOSName() == OSName.Unix ? "/src/main/deploy/" : "\\src\\main\\deploy\\";
        public static final String pathPlannerDir = getOSName() == OSName.Unix ? "pathplanner/" : "pathplanner\\";
        public static final String commandDirectory = getOSName() == OSName.Unix ? "/src/main/java/frc/robot/commands/" : "\\src\\main\\java\\frc\\robot\\commands\\";
        public static final String commandsRegex = getOSName() == OSName.Unix ? ".*/(.*).java" : ".*\\\\(.*).java";
        public static final String pathFileRegex = getOSName() == OSName.Unix ? ".*/(.*)\\.path" : ".*\\\\(.*)\\.path";
        public static final String autoModeAnnotationRegex = "@AutonomousModeAnnotation\\(parameterNames = \\{(.*)}\\)";

    }

        public enum OSName {
            Unix,
            Windows
        }
    }

File: app/src/main/java/java238/background/RobotProject.kt
/**
 * loads the frc robot project and exposes methods to get basic info
 */
package java238.background

import com.fasterxml.jackson.core.JsonProcessingException
import com.fasterxml.jackson.core.type.TypeReference
import com.fasterxml.jackson.databind.MapperFeature
import com.fasterxml.jackson.databind.ObjectMapper
import java.io.File
import java.io.FileNotFoundException
import java.io.IOException
import java.lang.reflect.Constructor
import java.net.URL
import java.net.URLClassLoader
import java.nio.file.Files
import java.nio.file.Paths
import java.util.Scanner
import java.util.regex.Pattern
import kotlin.collections.ArrayList

class RobotProject {
    private val mapper: ObjectMapper = ObjectMapper()
    private var jsonString: String? = null
    var trajectories: ArrayList<String>? = null
        private set

    @Throws(FileNotFoundException::class)
    fun indexCommands() {
        val pattern = Pattern.compile(Constants.autoModeAnnotationRegex)
        val dir = File(rootDirectory + Constants.commandDirectory)
        val directoryListing = dir.listFiles() ?: return
        for (child in directoryListing) {
            val info: CommandInfo
            val params: Array<String?>
            val scanner = Scanner(child)
            var line = scanner.nextLine()
            var commandConstructor : Constructor<*>? = null
            var foundAnnotation: Boolean = false
            var paramClasses = arrayListOf<Class<Any>>()
            var findPattern:String? = null
            // Do something with child
            while (scanner.hasNextLine()) {
                line = scanner.nextLine()
                findPattern = scanner.findInLine(pattern) ?: continue
                foundAnnotation = true
            }
            if (!foundAnnotation) continue

            val name = child.name.substring(0, child.name.length - 5)

            params = findPattern!!.substring(44, findPattern.length - 2).split(", ".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()

            val paramsList: ArrayList<String?> = params.toMutableList() as ArrayList<String?>
            val urls: Array<URL> = Array(4){File("$rootDirectory/build/classes/java/main/").toURI().toURL()}
            urls[1] =  File( "/home/haydenm/wpilib/2023/maven/edu/wpi/first/wpilibj/wpilibj-java/2023.4.2/wpilibj-java-2023.4.2.jar/").toURI().toURL()
            urls[2] =  File( "/home/haydenm/wpilib/2023/maven/edu/wpi/first/wpilibNewCommands/wpilibNewCommands-java/2023.4.2/wpilibNewCommands-java-2023.4.2.jar/").toURI().toURL()
            urls[3] =  File( "/home/haydenm/wpilib/2023/maven/edu/wpi/first/wpiutil/wpiutil-java/2023.4.2/wpiutil-java-2023.4.2.jar/").toURI().toURL()
            val loader: ClassLoader = URLClassLoader(urls)
            val loadedClass = loader.loadClass("frc.robot.commands.$name")
            for (constructor in loadedClass.constructors) {
                if (constructor.parameterCount == params.size) {
                    commandConstructor = constructor

                }
            }
            val parameterTypes: Array<Class<*>> = commandConstructor!!.parameterTypes
            info = CommandInfo(name, paramsList, parameterTypes)
            Companion.commands.add(info)
            println(info.name + " with parameters " + info.parameters.toString().strip())
        }
    }

    fun indexTrajectories() {
        trajectories = ArrayList()
        val pattern = Pattern.compile(rootDirectory + Constants.deployDirectory + Constants.pathPlannerDir)
        val dir = File(rootDirectory + Constants.deployDirectory + Constants.pathPlannerDir)
        val dirList = dir.listFiles()!!
        for (child in dirList) {
            println(child.name)
            trajectories!!.add(child.name.split("\\.path".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()[0])
        }
        println(trajectories)
    }

    init {
        mapper.configure(MapperFeature.ACCEPT_CASE_INSENSITIVE_PROPERTIES, true)
    }

    fun loadRobotProject() {
        jsonString = try {
            Files.readString(Paths.get(rootDirectory + Constants.deployDirectory + Constants.amodeFile))
        } catch (e: IOException) {
            throw RuntimeException(e)
        }
    }

    val commands: List<CommandInfo>
        get() = Companion.commands
    var autoModes: AmodeList?
        get() = if (amodesList != null) {
            amodesList
        } else {
            try {
                amodesList = mapper.readValue(jsonString, object : TypeReference<AmodeList?>() {})
            } catch (e: JsonProcessingException) {
                throw RuntimeException(e)
            }
            amodesList
        }
        set(list) {
            amodesList = list
            saveOut()
        }

    fun saveOut() {
        try {
            println(mapper.writeValueAsString(amodesList))
        } catch (e: JsonProcessingException) {
            throw RuntimeException(e)
        }
    }

    fun tempSaveDraggedCommand() {}

    object Constants {
        val amodeFile = "amode238.txt"
        val deployDirectory = if (oSName == OSName.Unix) "/src/main/deploy/" else "\\src\\main\\deploy\\"
        val pathPlannerDir = if (oSName == OSName.Unix) "pathplanner/" else "pathplanner\\"
        val commandDirectory = if (oSName == OSName.Unix) "/src/main/java/frc/robot/commands/" else "\\src\\main\\java\\u000crc\\robot\\commands\\"
        val commandsRegex = if (oSName == OSName.Unix) ".*/(.*).java" else ".*\\\\(.*).java"
        val pathFileRegex = if (oSName == OSName.Unix) ".*/(.*)\\.path" else ".*\\\\(.*)\\.path"
        val autoModeAnnotationRegex = "@AutonomousModeAnnotation\\(parameterNames = \\{(.*)}\\)"
    }

    enum class OSName {
        Unix, Windows
    }

    companion object {
        lateinit var rootDirectory: String
        private var amodesList: AmodeList? = null
        private val commands: MutableList<CommandInfo> = ArrayList()
        val oSName: OSName
            get() = if (System.getProperty("os.name").lowercase().contains("win")) OSName.Windows else OSName.Unix
    }

}